### ISSUE TITLE
calling a script

### ISSUE THREAD

**Issue by deblauwetom** [troglobit/watchdogd#14]
Hello,

Hereby a request for a feature, or maybe I don't get how to do it :) 

So I have a linux based platform (RPI3) and on that my main application is using the libwdog library to subscribe and kick2 the watchdog. So at startup of the platform, the app is not yet running, so the watchdog is not monitoring anything yet. However, I actually startup with doing some setup like for instance setting up an wifi AP, and also doing some fsck of the disks. So that is done in just a bash script. I am wondering how I could make that setup step secured by the watchdog?

Some other watchdog I used allows to give a script in the configuration that is periodically ran, and then with the exit code, it knows if something is wrong or not. Previously I used this feature to check if my main app was still not running after a certain amount of time after boot and if a certain timestamp of a file in /tmp was still being updated by the app. If not, then I considered it an error condition. 

So I don't see any means to do this with watchdogd itself I think? I could of course write a little app that subscribes to watchdogd and does this, but maybe something like this could be better added to watchdogd?

Best regards,

**Comment by troglobit** [comment]
Interesting idea. Had a similar question at work for generic monitor scripts.

I'll give it some thought and see what I can come up with. (No promises on a timeline though)

**Comment by deblauwetom** [comment]
Super! I also could help If you give Some hints How you would like it to be implemented!

**Comment by troglobit** [comment]
Dunno really, but one way of doing it could be to have an `/etc/watchdog.d/` directory where an operator could drop scripts to be run periodically in a run-parts sort of manner. Another could be to add a generic section to the .conf file, or support multiples .conf files in the above directory, something like this:

```
generic SOME_UNIQUE_NAME {
    interval = MSEC
    timeout = MSEC
    logmark = <true | false>
    warning = "Some warning message"
    critical = "Some critical message before reboot"
    script = /path/to/script
}
```

Runs script every interval, with an optional timeout to complete, optional warning and critical messages to log (if logmark is true).  If script returns non-zero => warning or critical, if script returns >1 => critical, or if it doesn't return => critical.
 
Not really sure where to take this yet. Have to spend some more time thinking about it. Possibly both alternatives could be supported.

**Comment by deblauwetom** [comment]
Nice ideas! I am more for putting it in the .conf as a section instead of the runparts thing, because that way you keep it together so easier to see what will happen. Also, maybe is this a bit over-designing for a first iteration of the improvement with all the options you propose? The essence is first to be able to run the script periodically, and when the scripts returns non-zero, then you force a reset, otherwise you keep going, and don't mind how long the script runs.  So the "interval" parameter and the "script" parameter are maybe, for now, the only really required options? That could be already a useable start, and then it could be expanded with the extras you propose.

**Comment by troglobit** [comment]
Yeah I just listed everything I could come up with, we could sure start with a subset of settings. But we may need some intelligence to protect from scripts going rouge. E.g., if a script doesn't exit before its next interval we should probably not start a new instance.

**Comment by deblauwetom** [comment]
Yes that is true.
The timeout value could be the same as the WDT timeout value by default.

**Comment by troglobit** [comment]
OK, so are you planning to do a PR?

**Comment by deblauwetom** [comment]
Only if you don't want to or don't have the time, I am willing to give it a try?

**Comment by troglobit** [comment]
Any contributions are welcome. Currently I'm focused on other things, so if you'd like to give it a go then please go ahead! I can help out by answering questions and review ideas/patches.

**Comment by deblauwetom** [comment]
Ok thanks, I'll try :)

I was checking out the config section, and it seems not much more trouble to implement some critical or warning config option. So I was thinking, instead of giving a string as warning or critical, maybe we should just use an int for warning and see that as a sort of "level". And so when the script returns some exit-code >= 1 then we check that with the warning and critical level tresholds? Also, we can already make a clear warning or critical message with the name of the script and the exit code.

**Comment by troglobit** [comment]
Sure, that sounds like a good idea. Whip something up and we can talk about it.  Just take care to not block watchdogd while a script runs, see src/script.c for the existing asynchronous framework.

**Comment by troglobit** [comment]
Closing, fixed in PR #15
